File: apps/sonar/composeApp/src/androidMain/kotlin/chat/bitchat/sonar/push/SonarPushRegistration.kt
package chat.bitchat.sonar.push

import android.content.Context
import android.net.Uri
import android.os.SystemClock
import android.util.Log
import chat.bitchat.sonar.AppContextHolder
import chat.bitchat.sonar.BuildConfig
import chat.bitchat.sonar.SonarCore
import chat.bitchat.sonar.wallet.WalletBridge
import chat.bitchat.sonar.wallet.WalletState
import com.google.firebase.messaging.FirebaseMessaging
import java.security.MessageDigest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout

/**
 * Registers the device's FCM token with both notification servers:
 *   1. Transponder — MIP-05 encrypted token shares (chat/call wakeups)
 *   2. Breez NDS — webhook URL (wallet wakeups, silent only)
 */
object SonarPushRegistration {

    private const val TAG = "SonarPush"
    private const val MAX_RETRIES = 3
    private const val DEFAULT_NDS_HOST = "nds.sonar.hedwig.sh"
    private const val WEBHOOK_MARKER_VERSION = "android-fcm-explicit-token-v2"
    /**
     * Persisted only for diagnostics (iOS `breez_webhook_marker` parity in
     * `SonarPushRegistration.swift`); NEVER used as a cross-launch skip — Boltz
     * owns the authoritative offer webhook state, and trusting a stale local
     * marker would suppress the unregister -> register self-heal. Updated in
     * place on success, removed on [unregister].
     */
    private const val WEBHOOK_MARKER_PREF_KEY = "breez_webhook_marker"
    private const val WEBHOOK_IN_FLIGHT_TIMEOUT_MS = 30_000L
    private const val WEBHOOK_REGISTRATION_TIMEOUT_MS = 20_000L
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val webhookLock = Any()

    // Same app-private "sonar" prefs used by WalletBridge.android.kt.
    private fun prefs() = AppContextHolder.ctx.getSharedPreferences("sonar", Context.MODE_PRIVATE)

    private val transponderNpub: String get() = BuildConfig.TRANSPONDER_NPUB

    // Base URL of the Breez NDS. Tolerates a bare host (prepends https://) and
    // falls back to the production NDS host for missing/truncated settings so a
    // malformed build setting cannot disable offline pay.
    private val ndsUrl: String
        get() {
            return normalizedNdsUrl(BuildConfig.NDS_URL)
        }

    @Volatile private var cachedFcmToken: String? = null
    @Volatile private var cachedOffer: String? = null
    @Volatile private var completedSessionWebhookMarker: String? = null
    @Volatile private var inFlightWebhookMarker: String? = null
    @Volatile private var inFlightWebhookStartedAtMs: Long = 0L
    @Volatile private var inFlightWebhookGeneration: Long = 0L

    fun ensureRegistered() {
        if (!SonarPushPrefs.effectivePushEnabled(AppContextHolder.ctx)) {
            Log.d(TAG, "Push not registered: disabled by user preference")
            return
        }
        // NEVER register (or start the core) before an account exists.
        // ensureRegistered() runs from SonarApp.onCreate; without this gate a
        // fresh install with default push prefs + a configured transponder
        // would reach SonarCore.start() → loadOrCreateIdentity() and MINT a new
        // nsec, which onboarding-complete inference then treats as a finished
        // account — booting an empty identity and skipping onboarding. Push has
        // nothing to register for until onboarding has produced a key.
        if (!SonarCore.hasIdentity()) {
            Log.d(TAG, "Push not registered: no account yet (pre-onboarding)")
            return
        }
        if (transponderNpub.isBlank() && ndsUrl.isBlank()) {
            Log.d(TAG, "Push not configured (no TRANSPONDER_NPUB or NDS_URL)")
            return
        }
        FirebaseMessaging.getInstance().token.addOnSuccessListener { token ->
            Log.d(TAG, "FCM token collected")
            cachedFcmToken = token
            registerTransponder(token)
            retryBreezWebhookIfNeeded()
        }.addOnFailureListener { e ->
            Log.w(TAG, "FCM token collection failed", e)
        }
    }

    fun onTokenRefresh(token: String) {
        if (!SonarPushPrefs.effectivePushEnabled(AppContextHolder.ctx)) {
            Log.d(TAG, "FCM token refreshed while push disabled")
            cachedFcmToken = null
            return
        }
        Log.d(TAG, "FCM token refreshed")
        cachedFcmToken = token
        registerTransponder(token)
        retryBreezWebhookIfNeeded()
    }

    /**
     * The Breez webhook is offer-scoped on the swap server. Registering only the
     * FCM token can leave an existing BOLT12 offer with a stale/missing webhook,
     * so coordinate token + current offer and force a re-subscribe when either
     * changes. Mirrors the iOS `unregisterWebhook` -> `registerWebhook` path,
     * without trusting a persisted local marker across app launches.
     */
    fun ensureBreezWebhook(offer: String) {
        cachedOffer = offer
        retryBreezWebhookIfNeeded()
    }

    /**
     * Retry Breez webhook registration after the wallet becomes Ready.
     * Called from the wallet setup path once the SDK connects.
     */
    fun retryBreezWebhookIfNeeded() {
        val token = cachedFcmToken
        val offer = cachedOffer
        if (token == null) {
            Log.d(TAG, "Breez NDS: receive offer ready, waiting for FCM token")
            return
        }
        if (offer == null) {
            Log.d(TAG, "Breez NDS: FCM token ready, waiting for receive offer")
            return
        }
        registerBreezWebhook(token, offer)
    }

    private fun registerTransponder(fcmToken: String) {
        if (transponderNpub.isBlank()) return
        // Guard the SonarCore.start() below: never mint an identity from a push
        // callback (token-refresh / background wakeup) before onboarding.
        if (!SonarCore.hasIdentity()) {
            Log.d(TAG, "Transponder: skipped, no account yet")
            return
        }
        scope.launch {
            var backoff = 2_000L
            for (attempt in 1..MAX_RETRIES) {
                try {
                    SonarCore.start()
                    SonarCore.connectRelays()
                    SonarCore.registerPushToken(
                        platform = "fcm",
                        token = fcmToken.toByteArray(Charsets.UTF_8),
                        serverNpub = transponderNpub,
                    )
                    Log.d(TAG, "Transponder: MIP-05 push token registered")
                    return@launch
                } catch (e: Exception) {
                    Log.w(TAG, "Transponder registration attempt $attempt/$MAX_RETRIES failed", e)
                    if (attempt < MAX_RETRIES) delay(backoff)
                    backoff *= 2
                }
            }
        }
    }

    private fun registerBreezWebhook(fcmToken: String, offer: String) {
        if (ndsUrl.isBlank()) return
        if (WalletBridge.state() !is WalletState.Ready) {
            Log.d(TAG, "Breez NDS: wallet not ready, will retry after wallet setup")
            return
        }
        val webhookUrl = webhookUrl(fcmToken)
        if (webhookUrl == null) {
            Log.w(TAG, "Breez NDS disabled: invalid NDS_URL build setting")
            return
        }
        val marker = webhookMarker(offer, webhookUrl)
        val attemptGeneration = synchronized(webhookLock) {
            if (completedSessionWebhookMarker == marker) {
                Log.d(TAG, "Breez NDS webhook already re-subscribed for current offer this launch")
                return
            }
            val now = SystemClock.elapsedRealtime()
            if (inFlightWebhookMarker == marker) {
                if (now - inFlightWebhookStartedAtMs < WEBHOOK_IN_FLIGHT_TIMEOUT_MS) {
                    Log.d(TAG, "Breez NDS webhook re-subscribe already in flight for current offer")
                    return
                }
                Log.w(TAG, "Breez NDS webhook re-subscribe timed out; retrying current offer")
            }
            inFlightWebhookMarker = marker
            inFlightWebhookStartedAtMs = now
            inFlightWebhookGeneration += 1
            inFlightWebhookGeneration
        }
        scope.launch {
            try {
                forceRegisterWebhook(webhookUrl)
                if (finishWebhookAttempt(marker, attemptGeneration, completed = true)) {
                    Log.d(TAG, "Breez NDS webhook force re-subscribed for current offer")
                }
            } catch (e: Exception) {
                if (finishWebhookAttempt(marker, attemptGeneration, completed = false)) {
                    Log.w(TAG, "Breez NDS webhook registration failed", e)
                }
            }
        }
    }

    private fun finishWebhookAttempt(
        marker: String,
        generation: Long,
        completed: Boolean,
    ): Boolean {
        return synchronized(webhookLock) {
            if (inFlightWebhookMarker != marker || inFlightWebhookGeneration != generation) {
                Log.d(TAG, "Breez NDS webhook ignoring stale re-subscribe attempt")
                return@synchronized false
            }
            inFlightWebhookMarker = null
            inFlightWebhookStartedAtMs = 0L
            if (completed) {
                completedSessionWebhookMarker = marker
                // Diagnostics-only persistence (update-in-place), mirroring iOS.
                // The in-memory session marker above stays the fast path.
                runCatching { prefs().edit().putString(WEBHOOK_MARKER_PREF_KEY, marker).apply() }
            }
            true
        }
    }

    fun unregister() {
        scope.launch {
            try { WalletBridge.unregisterWebhook() } catch (_: Exception) {}
        }
        cachedFcmToken = null
        cachedOffer = null
        synchronized(webhookLock) {
            completedSessionWebhookMarker = null
            inFlightWebhookMarker = null
            inFlightWebhookStartedAtMs = 0L
            inFlightWebhookGeneration += 1
        }
        // Force a fresh subscribe next time (e.g. after a wallet/seed change),
        // matching iOS `UserDefaults.removeObject(forKey: webhookMarkerKey)`.
        runCatching { prefs().edit().remove(WEBHOOK_MARKER_PREF_KEY).apply() }
        Log.d(TAG, "Unregistered from push servers")
    }

    /**
     * Identity replacement keeps the device FCM token but must forget every
     * offer-scoped marker from the previous deterministic wallet. Best-effort
     * unregister while the old node is still connected; the new wallet always
     * force-registers its own offer after setup.
     */
    suspend fun prepareForAccountReplacement() {
        try { WalletBridge.unregisterWebhook() } catch (_: Exception) {}
        cachedOffer = null
        synchronized(webhookLock) {
            completedSessionWebhookMarker = null
            inFlightWebhookMarker = null
            inFlightWebhookStartedAtMs = 0L
            inFlightWebhookGeneration += 1
        }
        runCatching {
            kotlinx.coroutines.withContext(Dispatchers.IO) {
                prefs().edit().remove(WEBHOOK_MARKER_PREF_KEY).commit()
            }
        }
    }

    internal fun normalizedNdsUrl(rawValue: String?): String {
        val raw = rawValue?.trim().orEmpty()
        if (raw.isEmpty() || raw == "https:" || raw == "http:") return "https://$DEFAULT_NDS_HOST"
        if (raw.startsWith("http://") || raw.startsWith("https://")) {
            val uri = runCatching { Uri.parse(raw) }.getOrNull()
            val host = uri?.host
            if (host.isNullOrBlank()) return "https://$DEFAULT_NDS_HOST"
            // Breez/Boltz requires HTTPS webhook URLs. Normalize any old
            // public `http://` setting instead of propagating it.
            return uri.buildUpon().scheme("https").build().toString()
        }
        return "https://$raw"
    }

    internal fun webhookUrl(fcmToken: String): String? {
        val uri = runCatching { Uri.parse(ndsUrl.trim()) }.getOrNull() ?: return null
        val scheme = uri.scheme?.lowercase()
        if (scheme != "https" || uri.host.isNullOrBlank()) return null
        return uri.buildUpon()
            .appendPath("api")
            .appendPath("v1")
            .appendPath("notify")
            .appendQueryParameter("platform", "android")
            .appendQueryParameter("token", fcmToken)
            .build()
            .toString()
    }

    private fun webhookMarker(offer: String, webhookUrl: String): String {
        val bytes = MessageDigest.getInstance("SHA-256")
            .digest("$WEBHOOK_MARKER_VERSION|$offer|$webhookUrl".toByteArray(Charsets.UTF_8))
        return bytes.joinToString(separator = "") { "%02x".format(it.toInt() and 0xff) }
    }

    private suspend fun forceRegisterWebhook(webhookUrl: String) {
        var lastError: Exception? = null
        for (attempt in 1..MAX_RETRIES) {
            try {
                withTimeout(WEBHOOK_REGISTRATION_TIMEOUT_MS) {
                    try { WalletBridge.unregisterWebhook() } catch (_: Exception) {}
                    WalletBridge.registerWebhook(webhookUrl)
                }
                return
            } catch (e: Exception) {
                lastError = e
                if (attempt < MAX_RETRIES) delay(attempt * 2_000L)
            }
        }
        throw lastError ?: IllegalStateException("webhook registration failed")
    }
}

File: ios/bitchat/Services/SonarPushRegistration.swift
//
// SonarPushRegistration.swift
// bitchat
//
// Registers the device's APNS token with both notification servers:
//   1. Transponder — MIP-05 encrypted token shares (chat/call wakeups)
//   2. Breez NDS   — webhook URL (wallet wakeups, silent only)
//
// This is free and unencumbered software released into the public domain.
// For more information, see <https://unlicense.org>
//

#if os(iOS)

import Foundation
import os
import CryptoKit
import SonarCore

/// Mutable registration state is confined to `queue`; async wallet work hops
/// back to that queue before reading or writing those fields.
final class SonarPushRegistration: @unchecked Sendable {
    static let shared = SonarPushRegistration()

    private static let log = Logger(
        subsystem: Bundle.main.bundleIdentifier ?? "sh.hedwig.sonar",
        category: "SonarPush"
    )

    private var cachedAPNSToken: Data?
    private var cachedFCMToken: String?
    private var cachedOffer: String?
    private var sonarNode: SonarNode?
    private let queue = DispatchQueue(label: "chat.bitchat.sonar.push.registration")

    /// Persisted only for diagnostics; do not use it as a cross-launch skip.
    /// Boltz owns the authoritative offer webhook state, and a stale local marker
    /// can otherwise suppress the unregister -> register self-heal.
    private static let webhookMarkerKey = "breez_webhook_marker"
    private static let webhookMarkerVersion = "ios-fcm-explicit-token-v2"
    private static let defaultNdsHost = "nds.sonar.hedwig.sh"
    private static let webhookInFlightTimeout: TimeInterval = 30
    private static let webhookRegistrationAttempts = 3
    private var completedSessionWebhookMarker: String?
    private var inFlightWebhookMarker: String?
    private var inFlightWebhookStartedAt: Date?
    private var inFlightWebhookGeneration: UInt64 = 0

    private var transponderNpub: String {
        Bundle.main.infoDictionary?["TRANSPONDER_NPUB"] as? String ?? ""
    }

    /// Base URL of the Breez NDS, e.g. `https://nds.sonar.hedwig.sh`.
    ///
    /// The value is configured in `Local.xcconfig` as a BARE HOST (no scheme),
    /// because xcconfig treats `//` as a comment and silently truncates
    /// `https://host` to `https:`. We prepend the scheme here and fall back to
    /// the production NDS host for missing/truncated settings so a malformed
    /// build setting does not disable offline receive in TestFlight/Release.
    private var ndsUrl: String {
        Self.normalizedNdsUrl(Bundle.main.infoDictionary?["NDS_URL"] as? String)
    }

    private init() {}

    // MARK: - Public

    func didRegisterForRemoteNotifications(deviceToken: Data) {
        let hex = deviceToken.map { String(format: "%02x", $0) }.joined()
        Self.log.info("APNS token collected (\(hex.prefix(8))...)")
        // Transponder (chat/calls) uses the raw APNs token. The Breez NDS uses the
        // Firebase FCM token instead — registered via didReceiveFCMToken.
        queue.async {
            self.cachedAPNSToken = deviceToken
            self.registerTransponderIfReady(token: deviceToken)
        }
    }

    /// Firebase FCM token for the Breez NDS webhook. `breez/notify` is FCM-only;
    /// Firebase bridges FCM → APNs on iOS (the APNs key lives in the Firebase
    /// Console). Matches Sonar Android and Unify.
    func didReceiveFCMToken(_ fcmToken: String, wallet: WalletBridgeService?) {
        Self.log.info("FCM token collected (\(fcmToken.prefix(8))...)")
        // Coordinate offer + token on the serial queue: the webhook can only be
        // (re)subscribed once both are known (the offer keys the idempotency
        // marker). Whichever lands second on the queue triggers the subscribe.
        queue.async {
            self.cachedFCMToken = fcmToken
            guard let offer = self.cachedOffer else {
                Self.log.info("Breez NDS: FCM token ready, waiting for receive offer")
                return
            }
            self.subscribeBreezWebhook(offer: offer, fcmToken: fcmToken, wallet: wallet)
        }
    }

    /// Called from the receiver-advertising flow each time the BOLT12 receive
    /// offer is (re)published. Coordinates the offer with the FCM token — whichever
    /// arrives second triggers the subscribe.
    func ensureBreezWebhook(offer: String, wallet: WalletBridgeService?) {
        queue.async {
            self.cachedOffer = offer
            guard let fcmToken = self.cachedFCMToken else {
                Self.log.info("Breez NDS: receive offer ready, waiting for FCM token")
                return
            }
            self.subscribeBreezWebhook(offer: offer, fcmToken: fcmToken, wallet: wallet)
        }
    }

    func retryBreezWebhookIfNeeded(wallet: WalletBridgeService) {
        queue.async {
            guard let fcmToken = self.cachedFCMToken, let offer = self.cachedOffer else { return }
            self.subscribeBreezWebhook(offer: offer, fcmToken: fcmToken, wallet: wallet)
        }
    }

    func unregister(wallet: WalletBridgeService? = nil) {
        queue.async {
            self.cachedAPNSToken = nil
            self.cachedOffer = nil
            self.completedSessionWebhookMarker = nil
            self.inFlightWebhookMarker = nil
            self.inFlightWebhookStartedAt = nil
            self.inFlightWebhookGeneration &+= 1
            self.sonarNode = nil
        }
        // Force a fresh subscribe next time (e.g. after a wallet/seed change).
        UserDefaults.standard.removeObject(forKey: Self.webhookMarkerKey)
        if let wallet {
            Task { @MainActor in
                try? await wallet.unregisterWebhook()
            }
        }
        Self.log.info("Unregistered from push servers")
    }

    /// Account replacement keeps the device APNs/FCM tokens but invalidates the
    /// previous wallet's offer-scoped Breez markers. Unregister the old offer
    /// while its wallet is still connected; the replacement wallet can then
    /// register immediately without waiting for another token callback.
    ///
    /// Do not clear `sonarNode` here: wallet cleanup can still fail before the
    /// identity is mutated, in which case the old account must keep receiving
    /// push registrations. `MarmotService.wipeDatabase()` clears the node at the
    /// actual identity-replacement commit point.
    func prepareForAccountReplacement(wallet: WalletBridgeService?) async {
        if let wallet {
            try? await wallet.unregisterWebhook()
        }
        queue.sync {
            cachedOffer = nil
            completedSessionWebhookMarker = nil
            inFlightWebhookMarker = nil
            inFlightWebhookStartedAt = nil
            inFlightWebhookGeneration &+= 1
        }
        UserDefaults.standard.removeObject(forKey: Self.webhookMarkerKey)
    }

    /// Set the SonarNode once it's initialized. Retries transponder registration
    /// if an APNS token was already collected.
    func setSonarNode(_ node: SonarNode) {
        queue.async {
            self.sonarNode = node
            if let token = self.cachedAPNSToken {
                self.registerTransponderIfReady(token: token)
            }
        }
    }

    /// Release the core node when the owning Marmot session is erased. The core
    /// cache generation guard also rejects late writes, but retaining a wiped
    /// session here would keep its runtime and other sensitive state alive.
    func clearSonarNode() {
        queue.sync {
            self.sonarNode = nil
        }
    }

    // MARK: - Private

    private static let maxRetries = 3

    private func registerTransponderIfReady(token: Data) {
        guard !transponderNpub.isEmpty else { return }
        guard let node = sonarNode else {
            Self.log.info("Transponder: SonarNode not ready, will retry after setSonarNode")
            return
        }
        let npub = self.transponderNpub
        DispatchQueue.global(qos: .utility).async {
            var backoff: UInt32 = 2
            for attempt in 1...Self.maxRetries {
                do {
                    try node.registerPushToken(
                        platform: "apns",
                        token: token,
                        serverNpub: npub
                    )
                    Self.log.info("Transponder: MIP-05 push token registered")
                    return
                } catch {
                    Self.log.warning("Transponder registration attempt \(attempt)/\(Self.maxRetries) failed: \(error)")
                    if attempt < Self.maxRetries { sleep(backoff) }
                    backoff *= 2
                }
            }
        }
    }

    private func subscribeBreezWebhook(offer: String, fcmToken: String, wallet: WalletBridgeService?) {
        guard let webhookUrl = Self.webhookUrl(ndsUrl: ndsUrl, platform: "ios", fcmToken: fcmToken) else {
            Self.log.warning("Breez NDS disabled: invalid NDS_URL build setting")
            return
        }
        guard let wallet else {
            Self.log.info("Breez NDS: wallet not ready, will retry after wallet setup")
            return
        }
        let marker = Self.webhookMarker(offer: offer, webhookUrl: webhookUrl)
        if completedSessionWebhookMarker == marker {
            Self.log.info("Breez NDS webhook already re-subscribed for current offer this launch")
            return
        }
        if inFlightWebhookMarker == marker {
            if let started = inFlightWebhookStartedAt,
               Date().timeIntervalSince(started) < Self.webhookInFlightTimeout {
                Self.log.info("Breez NDS webhook re-subscribe already in flight for current offer")
                return
            }
            Self.log.warning("Breez NDS webhook re-subscribe timed out; retrying current offer")
        }
        inFlightWebhookMarker = marker
        inFlightWebhookStartedAt = Date()
        inFlightWebhookGeneration &+= 1
        let attemptGeneration = inFlightWebhookGeneration
        Task { @MainActor in
            guard case .ready = wallet.state else {
                Self.log.info("Breez NDS: wallet not ready, will retry after wallet setup")
                self.finishWebhookAttempt(marker: marker, generation: attemptGeneration, completed: false)
                return
            }
            do {
                try await Self.forceRegisterWebhook(wallet: wallet, url: webhookUrl)
                self.finishWebhookAttempt(marker: marker, generation: attemptGeneration, completed: true)
            } catch {
                self.finishWebhookAttempt(
                    marker: marker,
                    generation: attemptGeneration,
                    completed: false,
                    failureMessage: "\(error)"
                )
            }
        }
    }

    private func finishWebhookAttempt(
        marker: String,
        generation: UInt64,
        completed: Bool,
        failureMessage: String? = nil
    ) {
        queue.async {
            guard self.inFlightWebhookMarker == marker,
                  self.inFlightWebhookGeneration == generation
            else {
                Self.log.info("Breez NDS webhook ignoring stale re-subscribe attempt")
                return
            }
            self.inFlightWebhookMarker = nil
            self.inFlightWebhookStartedAt = nil
            if completed {
                UserDefaults.standard.set(marker, forKey: Self.webhookMarkerKey)
                self.completedSessionWebhookMarker = marker
                Self.log.info("Breez NDS webhook force re-subscribed for current offer (FCM)")
            } else if let failureMessage {
                Self.log.warning("Breez NDS webhook registration failed: \(failureMessage)")
            }
        }
    }

    static func normalizedNdsUrl(_ rawValue: String?) -> String {
        let raw = (rawValue ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        guard !raw.isEmpty, raw != "https:", raw != "http:" else {
            return "https://\(defaultNdsHost)"
        }
        if raw.hasPrefix("http://") || raw.hasPrefix("https://") {
            guard var components = URLComponents(string: raw),
                  components.host?.isEmpty == false
            else {
                return "https://\(defaultNdsHost)"
            }
            // Breez/Boltz requires HTTPS webhook URLs. Normalize any old
            // TestFlight/public `http://` setting instead of propagating it.
            components.scheme = "https"
            return components.url?.absoluteString ?? "https://\(defaultNdsHost)"
        }
        return "https://\(raw)"
    }

    static func webhookUrl(ndsUrl: String, platform: String, fcmToken: String) -> String? {
        let raw = ndsUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        guard var components = URLComponents(string: raw),
              let scheme = components.scheme?.lowercased(),
              scheme == "https",
              components.host?.isEmpty == false
        else { return nil }

        var path = components.path
        if path.hasSuffix("/") {
            path.removeLast()
        }
        components.path = "\(path)/api/v1/notify"
        var queryItems = components.queryItems ?? []
        queryItems.append(URLQueryItem(name: "platform", value: platform))
        queryItems.append(URLQueryItem(name: "token", value: fcmToken))
        components.queryItems = queryItems
        return components.url?.absoluteString
    }

    private static func webhookMarker(offer: String, webhookUrl: String) -> String {
        let digest = SHA256.hash(data: Data("\(webhookMarkerVersion)|\(offer)|\(webhookUrl)".utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    @MainActor
    private static func forceRegisterWebhook(wallet: WalletBridgeService, url: String) async throws {
        var lastError: Error?
        for attempt in 1...webhookRegistrationAttempts {
            do {
                // Force a clean re-subscribe. A plain registerWebhook no-ops when the
                // SDK's local `bolt12_offers.webhook_url` already equals this URL but
                // Boltz still holds `url=None` — the desync that breaks offline receive.
                // unregister clears the local webhook state, register re-PATCHes the
                // offer on Boltz (`update_bolt12_offer`) with the current FCM URL.
                // Matches the misty-breez reference's unregister→register flow.
                try? await wallet.unregisterWebhook()
                try await wallet.registerWebhook(url: url)
                return
            } catch {
                lastError = error
                guard attempt < webhookRegistrationAttempts else { break }
                try? await Task.sleep(nanoseconds: UInt64(attempt) * 2_000_000_000)
            }
        }
        throw lastError ?? WalletBridgeService.WalletBridgeError.core("webhook registration failed")
    }
}

#endif
